fxa-oauth-server/lib/routes/introspect.js

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*jshint camelcase: false*/
+const Joi = require('joi');
+const db = require('../db');
+const encrypt = require('../encrypt');
+const validators = require('../validators');
+const hex = require('buf').to.hex;
+
+const PAYLOAD_SCHEMA = Joi.object({
+  token: Joi.string().required(),
+  token_type_hint: Joi.string().equal(['access_token', 'refresh_token'])
+});
+
+// The "token introspection" endpoint, per https://tools.ietf.org/html/rfc7662
+
+module.exports = {
+  validate: {
+    payload: PAYLOAD_SCHEMA.options({ stripUnknown: true })
+  },
+  response: {
+    schema: Joi.object().keys({
+      // https://tools.ietf.org/html/rfc7662#section-2.2
+      active: Joi.boolean().required(),
+      scope: validators.scope.optional(),
+      client_id: validators.clientId.optional(),
+      token_type: Joi.string().equal(['access_token', 'refresh_token']),
+      exp: Joi.number().optional(),
+      iat: Joi.number().optional(),
+      sub: Joi.string().optional(),
+      iss: Joi.string().optional(),
+      jti: Joi.string().optional(),
+      'fxa-lastUsedAt': Joi.number().optional()
+    })
+  },
+  handler: async function tokenEndpoint(req) {
+    let token;
+    let tokenType = req.payload.token_type;
+    const tokenId = encrypt.hash(req.payload.token);
+    if (tokenType === 'access_token' || ! tokenType) {
+      token = await db.getAccessToken(tokenId);
+      if (token) {
+        tokenType = 'access_token';
+      }
+    }
+    if (tokenType === 'refresh_token' || (! tokenType && ! token)) {
+      // XXX TODO: if the refreshToken belongs to a non-public client,
+      // we should require the client_secret in Authentication header.
```

Comment:
For the initial version, we could also just error out with an "unsupported" error of some kind, and do the extra check in a follow-up.

Reply:
sounds good 👍